lib/auth/auth_chalresp.c

```diff
@@ -0,0 +1,637 @@
+
+		/* canceled ? */
+		if (auth_conn->cancel_auth) {
+			*status = AUTH_STATUS_CANCELED;
+			return false;
+		}
+
+		/* timed out, try to read again */
+		if (numbytes == -EAGAIN) {
+			continue;
+		}
+
+		if (numbytes <= 0) {
+			LOG_ERR("Failed to read server challenge response, err: %d", numbytes);
+			*status = AUTH_STATUS_FAILED;
+			return false;
+		}
+
+		buf += numbytes;
+		len -= numbytes;
+	}
+
+	/* check message */
+	if (!auth_check_msg(&server_resp.hdr, AUTH_SERVER_CHALRESP_MSG_ID)) {
+		LOG_ERR("Invalid message received from the server.");
+		*status = AUTH_STATUS_FAILED;
+		return false;
+	}
+
+
+	/* Now verify response, is the response correct?  Hash the random challenge
+	 * with the shared key. */
+	err = auth_chalresp_hash(random_chal, hash);
+
+	if (err) {
+		LOG_ERR("Failed to calc hash, err: %d", err);
+		*status = AUTH_STATUS_FAILED;
+		return false;
+	}
+
+	/* Does the response match what is expected? */
+	if (memcmp(hash, server_resp.server_response, sizeof(hash))) {
+		/* authentication failed */
+		LOG_ERR("Server authentication failed.");
+		*status = AUTH_STATUS_AUTHENTICATION_FAILED;
+
+		/* send failed message to the Peripheral */
+		memset(&chal_result, 0, sizeof(chal_result));
+		chal_result.hdr.soh = CHALLENGE_RESP_SOH;
+		chal_result.hdr.msg_id = AUTH_CHALRESP_RESULT_MSG_ID;
+		chal_result.result = 1;
+
+		numbytes = auth_xport_send(auth_conn->xport_hdl, (uint8_t *)&chal_result, sizeof(chal_result));
+
+		if ((numbytes <= 0) || (numbytes != sizeof(chal_result))) {
+			LOG_ERR("Failed to send authentication error result to server.");
+		}
+
+		return false;
+	}
+
+	/* init Client response message */
+	memset(&client_resp, 0, sizeof(client_resp));
+	client_resp.hdr.soh = CHALLENGE_RESP_SOH;
+	client_resp.hdr.msg_id = AUTH_CLIENT_CHALRESP_MSG_ID;
+
+	/* Create response to the server's random challenge */
+	err = auth_chalresp_hash(server_resp.server_challenge, client_resp.client_response);
+
+	if (err) {
+		LOG_ERR("Failed to create server response to challenge, err: %d", err);
+		*status = AUTH_STATUS_FAILED;
+		return false;
+	}
+
+	/* send Client's response to the Server's random challenge */
+	numbytes = auth_xport_send(auth_conn->xport_hdl, (uint8_t *)&client_resp, sizeof(client_resp));
+
+	if ((numbytes <= 0) || (numbytes != sizeof(client_resp))) {
+		LOG_ERR("Failed to send Client response.");
+		*status = AUTH_STATUS_FAILED;
+		return false;
+	}
+
+	/* so far so good, need to wait for Server response */
+	*status = AUTH_STATUS_IN_PROCESS;
+	return true;
+}
+
+/**
+ * Server waits and receives a message from the client.
+ *
+ * @param auth_conn   Authentication connection structure.
+ * @param msgbuf      Buffer to copy message into.
+ * @param msglen      Buffer byte length.
+ *
+ * @return true if number of bytes were received, else false.
+ */
+static bool auth_server_recv_msg(struct authenticate_conn *auth_conn, uint8_t *msgbuf, size_t msglen)
+{
+	int numbytes;
+
+	while ((int)msglen > 0) {
```

Comment:
this is cast is wrong, if `msgelen` is smaller than `0` it means an overflow in `msgbuf`, if it is `0` you don't need to cast it to signed type. Inspect the code it does not seem  possible `msglen` be lesser than 0.

Another problem is that this cast can result in `< 0` for a big `msglen`.

Reply:
Fixed.